protoreflect: add shorthands for names of common persisted messages

dt opened the pull request:
This adds a mechanism for registering shorthand aliases for the names of
messages that are most often persisted including∑ jobspb.Progress,
jobspb.Payload, descpb.Descriptor, etc.

These shorthands are then consulted when looking up a message type by
name, such as when using pb_to_json or json_to_pb, if the provided name
is not found in the existing global registry that contains only the
fully qualified (by proto import path) names. Shorthands are folded to
lower case and any '_'s are removed for search. Thus, with this change,
one can use `crdb_internal.pb_to_json('desc', descriptor)` in place of
`crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', descriptor)`.

Release note: none.

dt committed:
protoreflect: add shorthands for names of common persisted messages

This adds a mechanism for registering shorthand aliases for the names of
messages that are most often persisted including∑ jobspb.Progress,
jobspb.Payload, descpb.Descriptor, etc.

These shorthands are then consulted when looking up a message type by
name, such as when using pb_to_json or json_to_pb, if the provided name
is not found in the existing global registry that contains only the
fully qualified (by proto import path) names. Shorthands are folded to
lower case and any '_'s are removed for search. Thus, with this change,
one can use `crdb_internal.pb_to_json('desc', descriptor)` in place of
`crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', descriptor)`.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74284)
<!-- Reviewable:end -->

miretskiy reviewed (approved):
> Reviewed 3 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74284)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @dt)

___
*[pkg/sql/protoreflect/utils.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74284#-MrnPPtAASFiLKj5w0iZ:-MrnPPtB1gf__eyUqBeE:b3nxf6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d431354cff18c8bccd12d2b5e87ebaf2b3aff926/pkg/sql/protoreflect/utils.go#L31)):*
> ```Go
> // Aliases are converted to and searched for using all lower-case.
> func RegisterShorthand(name string, msg protoutil.Message) {
> 	shorthands[strings.ToLower(name)] = msg
> ```

I think it's appropriate to panic if name already exists as a shorthand.


<!-- Sent from Reviewable.io -->

miretskiy commented:
I've been thinking: what If instead of doing the registration mechanism we attempt to look up based on the name and if it's not found we iterate over the list of all registered protocol messages and we see if any of them ends with whatever was specified.

dt commented:
> instead of doing the registration mechanism we attempt to look up based on the name and if it's not found we iterate over the list

I actually started by doing that (see thread on slack): I had to add an accessory to gogo's registry to get all the names but then on lookup failure I iterated the whole list and iff exactly one item contained the provided name I used it instead (if more than one message type name contained the provided name, I'd wrap the not found error explaining the ambiguity with the first two matches).

I was convinced there that the extra complexity -- and required upstream Gogo change -- just wasn't worth it when there are really only a couple heavily used message types.

miretskiy commented:
Agree

On Mon, Dec 27, 2021, 11:42 AM David Taylor ***@***.***>
wrote:

> instead of doing the registration mechanism we attempt to look up based on
> the name and if it's not found we iterate over the list
>
> I actually started by doing that (see thread on slack): I had to add an
> accessory to gogo's registry to get all the names but then on lookup
> failure I iterated the whole list and iff exactly one item contained the
> provided name I used it instead (if more than one message type name
> contained the provided name, I'd wrap the not found error explaining the
> ambiguity with the first two matches).
>
> I was convinced there that the extra complexity -- and required upstream
> Gogo change -- just wasn't worth it when there are really only a couple
> heavily used message types.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/74284#issuecomment-1001650131>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVGYDK56MTVX2TBNYKTUTCJOZANCNFSM5KYCW4PQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

postamar commented on pkg/sql/catalog/descpb/structured.go:
```diff
@@ -391,3 +392,9 @@ func (ni NameInfo) GetParentSchemaID() ID {
+func init() {
+	protoreflect.RegisterShorthand("Descriptor", (*Descriptor)(nil))
+	protoreflect.RegisterShorthand("desc", (*Descriptor)(nil))
+	protoreflect.RegisterShorthand("TableDescriptor", (*TableDescriptor)(nil))
```
Would `TableDescriptor` ever be used in practice? For what it's worth, I never have, it's always the `Descriptor` union type.

dt replied:
Good point, removed it.

postamar reviewed (commented):
Nice!

Nit: could you augment your commit with at least one actual usage of a shorthand, please? Here could be suitable, for instance: https://github.com/cockroachdb/cockroach/blob/7780e1e085dbfa849defcef63b5138e80c18a17b/pkg/sql/tests/repair_test.go#L251-L253

dt commented:
Added a example to commit, made duplicate registration panic as suggested, and added folding away of `_` too now so I don't need to remember if it is `ScheduleDetails` or `schedule_details` or whatever, and made is variadic for registering multiple aliases in one call, so we didn't need to register descriptor and desc separately.

dt commented:
TFTR!

bors r+

dt commented:
bors r+
